### PR TITLE
Bug - double ParBlockNonlinearForm::GetEnergy(const Vector &x)

### PR DESCRIPTION
**Pull request by bslazarov**
GetEnergy method was not implemented for ParBlockNonlinearForm. A call to GetEnergy was redirected to BlockNonlinearForm returning only the local process contributions. The current implementation returns the sum the contributions and return the result. 

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1702](https://github.com/mfem/mfem/pull/1702) | @bslazarov | @tzanio | @jandrej + @rcarson3 | 08/13/20 | 10/04/20 | 10/08/20 | |

### CODE REVIEW

File: fem/pnonlinearform.cpp [hunk 1]
```diff
@@ -217,6 +217,28 @@ void ParBlockNonlinearForm::SetEssentialBC(const
    }
 }
 
+double ParBlockNonlinearForm::GetEnergy(const Vector &x) const
+{
+
```
Comment (jandrej) [inline fem/pnonlinearform.cpp:222]: Empty line. Shouldn't astyle enforce that?

File: fem/pnonlinearform.cpp [hunk 2]
```diff
@@ -217,6 +217,28 @@ void ParBlockNonlinearForm::SetEssentialBC(const
...
+   xs_true.Update(x.GetData(), block_trueOffsets);
+   xs.Update(block_offsets);
+
+   for (int s=0; s<fes.Size(); ++s)
+   {
+      fes[s]->GetProlongationMatrix()->Mult(
+         xs_true.GetBlock(s), xs.GetBlock(s));
+   }
+
```
Comment (rcarson3) [inline fem/pnonlinearform.cpp:231]: ```ParNonlinearForm::GetEnergy``` has the following lines which should probably be added here as well:
```c++
   if (fnfi.Size())
   {
      MFEM_ABORT("TODO: add energy contribution from shared faces");
   }
```

Comment (bslazarov) [reply fem/pnonlinearform.cpp:231]: GetEnergy calls GetEnergyBlocked which has all these checks.

File: fem/pnonlinearform.cpp (hunk 1 above)
Comment (bslazarov) [reply fem/pnonlinearform.cpp:222]: Enforced by hand

Comment (bslazarov) [reply fem/pnonlinearform.cpp]: Unit test added!

File: fem/pnonlinearform.cpp (hunk 2 above)
Comment (bslazarov) [reply fem/pnonlinearform.cpp:230]: Unit test added!

**Comment by bslazarov** [comment]
The PR should be ready for merging.

**Comment by tzanio** [comment]
> The PR should be ready for merging.

It should pass Travis first 😝

**Comment by bslazarov** [comment]
I am not sure what is the problem with Travis. 

⁣Sent from Blue ​

**Comment by tzanio** [comment]
> /home/travis/build/mfem/mfem/tests/unit/fem/test_3d_parblocknlform.cpp: In member function ‘virtual double blocknonlinearform_3d::CExample::GetElementEnergy(const mfem::Array<const mfem::FiniteElement*>&, mfem::ElementTransformation&, const mfem::Array<const mfem::Vector*>&)’:
/home/travis/build/mfem/mfem/tests/unit/fem/test_3d_parblocknlform.cpp:60:11: error: unused variable ‘dim’ [-Werror=unused-variable]
       int dim = el[0]->GetDim();
           ^~~
cc1plus: all warnings being treated as errors
makefile:84: recipe for target 'fem/test_3d_parblocknlform.o' failed
make[1]: *** [fem/test_3d_parblocknlform.o] Error 1
make[1]: Leaving directory '/home/travis/build/mfem/mfem/tests/unit'
makefile:430: recipe for target 'tests/unit' failed

**Comment by bslazarov** [comment]
Now is ready :-)

⁣Sent from Blue ​

**Comment by jandrej** [comment]
I had to make quite a few changes to make the test reasonable. I also reformatted the files to be more readable.

**Comment by jandrej** [comment]
LGTM

**Review by jandrej — APPROVED** [review]

**Comment by tzanio** [comment]
@adrienbernede, @jandrej, do you understand the GitLab error?

**Comment by tzanio** [comment]
@rcarson3, can you take another look?

**Comment by jandrej** [comment]
LGTM

**Review by rcarson3 — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Review by tzanio — APPROVED** [review]
Thanks for the bug fix @bslazarov!

**Pull request merged**